Campaign observations - Abstract Point and Image data metadata

jomey opened the pull request:
This is quite a big one and refactors the structure around metadata for Point and Image data as discussed in #137 and closes related issues (bottom of this description)

It also updates the test suite to use [factory boy](https://factoryboy.readthedocs.io/en/stable/) along with the [integration into ptytest](https://pytest-factoryboy.readthedocs.io/en/stable/).

I will highlight more major changes in code comments.

Grab yourself a :coffee: (or :beer:) this PR has a lot to unfold.


Close #137
Close #136 
Close #132

jomey commented on snowexsql/api.py:
```diff
@@ -207,6 +230,103 @@ def from_unique_entries(cls, columns_to_search, **kwargs):
         return results
 
+    @classmethod
+    def from_filter(cls, **kwargs):
```
Moved this filter to the BaseClass as it seems more appropriate here.

jomey commented on snowexsql/api.py:
```diff
@@ -207,6 +230,103 @@ def from_unique_entries(cls, columns_to_search, **kwargs):
+                        compile_kwargs={"literal_binds": True}
+                    )
+                    print("\n ** SQL query **")
+                    print(full_sql_query)
```
Added a little debug method that can be triggered in a test suite

jomey commented on snowexsql/conversions.py:
```diff
@@ -54,8 +53,10 @@ def query_to_geopandas(query, engine, **kwargs):
+    # Get dataframe from geopandas using the query and the DB connection.
+    # By passing in the actual connection, we maintain ownership of it and
+    # keep it alive until we close it.
+    df = gpd.read_postgis(sql, engine.connect(), **kwargs)
```
This change to pass in the connection was key so we maintain ownership and pandas/geopandas won't close this on us.

jomey replied:
Using a `with` block would close the connection for us again and require another monkeypatch for the test suite. Let's do that in a follow up PR

jomey commented on snowexsql/tables/image_data.py:
```diff
@@ -1,20 +1,13 @@
     """
     __tablename__ = 'images'
-    # Date of the measurement
-    date = Column(Date)
```
Moved this to the observation

jomey commented on snowexsql/tables/measurement_type.py:
```diff
@@ -15,17 +15,17 @@ class MeasurementType(Base):
 
     @declared_attr
-    def measurement(cls):
+    def measurement_type(self):
```
This was a leftover from a PR that renamed all mentions of measurement to measurement_type

jomey commented on snowexsql/tables/point_data.py:
```diff
@@ -41,8 +24,16 @@ class PointData(
+        """
+        Helper attribute to only query for dates of measurements
+        """
+        return self.date.date()
```
Pretty cool SQLAlchemy feature:
https://docs.sqlalchemy.org/en/20/orm/extensions/hybrid.html#module-sqlalchemy.ext.hybrid

jomey commented on tests/conftest.py:
```diff
@@ -0,0 +1,100 @@
+# It will print the query from the BaseDataset.from_filter() method
+@pytest.fixture(scope='session')
+def _debug_sql_query():
+    os.environ['DEBUG_QUERY'] = '1'
```
Fixture when used that prints out the SQL query

jomey commented on tests/conftest.py:
```diff
@@ -0,0 +1,100 @@
+    def db_session(*args, **kwargs):
+        yield SESSION(), sqlalchemy_engine
+
+    monkeypatch.setattr(snowexsql.api, "db_session", db_session)
```
This forces the session created within the API class to use the test session. It also voids the need to overwrite the DB connection string and enables the use of transactions to roll back created test data with a test function

jomey commented on tests/db_connection.py:
This file should eventually be deleted once we have everything switched to using FactoryBoy

jomey commented on tests/db_setup.py:
```diff
@@ -1,16 +1,26 @@
+# DB Configuration and Session
+CREDENTIAL_FILE = join(dirname(__file__), 'credentials.json')
+DB_INFO = json.load(open(CREDENTIAL_FILE))
+SESSION = orm.scoped_session(orm.sessionmaker())
```
Very important to use a scoped_session
https://docs.sqlalchemy.org/en/20/orm/contextual.html#contextual-thread-local-sessions

jomey commented on tests/factories/point_observation.py:
```diff
@@ -0,0 +1,26 @@
+    doi = factory.SubFactory(DOIFactory)
+    instrument = factory.SubFactory(InstrumentFactory)
+    measurement_type = factory.SubFactory(MeasurementTypeFactory)
+    observer = factory.SubFactory(ObserverFactory)
```
One of my favorite features of factory boy to just link to other factory classes and it will take care of creating all the data

jomey commented on tests/test_db.py:
This should eventually be replaced by the table level tests in `tests/tables/`

jomey commented:
I am out of ideas why tests for Python 3.9 through 3.12 are hanging with GitHub actions.
Maybe @micah-prime or @aaarendt could check out this PR and run it locally.
Of course, all tests run fine with my local setup and Python 3.12

jomey commented:
> I am out of ideas why tests for Python 3.9 through 3.12 are hanging with GitHub actions. Maybe @micah-prime or @aaarendt could check out this PR and run it locally. Of course, all tests run fine with my local setup and Python 3.12

Well now also the tests for 3.8 are hanging and I only added a test case that checks for more information in the connection string.

When I shutdown my DB locally, the tests start failing fairly quickly. Makes me think it is not a DB connection issue

aaarendt commented:
The tests hang for me when it gets to test_db.py. The api and tables tests succeed. 

I'm looking at the content of ```pg_stat_activity``` while the test hangs and see this:

![image](https://github.com/user-attachments/assets/b1247d7a-9fc7-4fa0-8e14-9bc499c10c58)

Wondering if it's something related to postgresql being strict about locking concurrent transactions, and your new use of rolling back after each test? Maybe useful: [https://github.com/sqlalchemy/sqlalchemy/issues/4685](https://github.com/sqlalchemy/sqlalchemy/issues/4685)

jomey commented:
> The tests hang for me when it gets to test_db.py. The api and tables tests succeed.
> 
> I'm looking at the content of `pg_stat_activity` while the test hangs and see this:
> 
> Wondering if it's something related to postgresql being strict about locking concurrent transactions, and your new use of rolling back after each test? Maybe useful: [sqlalchemy/sqlalchemy#4685](https://github.com/sqlalchemy/sqlalchemy/issues/4685)

Thanks for the great detective work @aaarendt 
I have updated the test suite and removed the rollback with the old session setup. I am trying not to overhaul the whole test suite with this PR and do this in a separate effort. Might have to bite the bullet though and at least replace the tests in `test_db_tables` with the new structure.

aaarendt commented:
For my own learning, is it correct that e.g. lines  37-39 in api/test_layer_measurements.py (with reference to "fakeinstrument") is an example of old tests that will be superseded by your factory boy approach? 
(weird that you can't comment on code that hasn't been changed in a commit...?)

aaarendt reviewed (approved):
All the tests work locally when I switch to builder username. I've poked around the rest of the code to try to learn things and don't know enough yet to do a really rigorous review.

jomey commented:
> For my own learning, is it correct that e.g. lines 37-39 in api/test_layer_measurements.py (with reference to "fakeinstrument") is an example of old tests that will be superseded by your factory boy approach? (weird that you can't comment on code that hasn't been changed in a commit...?)

Correct. Just logged an issue for that #158

jomey commented on tests/test_db.py:
```diff
@@ -1,98 +1,86 @@
+class TestDBConnectionInfo:
+    def test_load_credentials(self):
+        user, password = load_credentials(DBSetup.CREDENTIAL_FILE)
+        assert user == 'builder'
```
Good call for not a good style.

Will update to compare against the value from the `credentials.json` just like on line 50 in this file

jomey commented on tests/factories/layer_data.py:
```diff
@@ -0,0 +1,28 @@
+    comments = 'Layer comment'
+    value = '40'
+    flags = 'Sample'
+    sample_a = '42'
```
@micah-prime - Can you remind me, why the `value` `and sample_(a|b|c)` columns are strings and not int column types?

micah-prime replied:
We're storing the Grain Type, Hand Hardness, and Manual Wetness data in the value column as well. I'm assuming the a, b, and c samples are for consistency with that data type

micah-prime reviewed (commented):
Nice work! Just a couple comments/questions

jomey commented on snowexsql/tables/point_data.py:
```diff
@@ -1,38 +1,21 @@
-    # Date of the measurement
-    date = Column(Date)
+    # Date of the measurement with time
+    date = Column(DateTime)
```
Very good point. I will rename to `datetime`. Also made me realize that I need to add `timezone=True` to all the new date/time definitions.

jomey commented on snowexsql/api.py:
```diff
@@ -287,98 +407,78 @@ class PointMeasurements(BaseDataset):
-                session.close()
-                LOG.error("Failed query for PointData")
-                raise e
+    def _filter_campaign(cls, qry, value):
```
I am leaning towards a new issue that takes a second look at the API and how we construct queries. Think with the new structure, we can sometimes use `joins` and sometimes [subquery expressions](https://www.postgresql.org/docs/current/functions-subquery.html).

What do you think?

micah-prime replied:
If that will simplify the queries and/or make the code more maintainable then I am for it!

jomey replied:
Logged #160
